### PR TITLE
[2.3] Static Code Analysis for Components

### PR DESCRIPTION
**Pull request by kalessil**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

Static Code Analysis with Php Inspections (EA Extended):
    - magic methods validity: resolved possible fatals
    - random api migration: used mt_\* in some of components
    - *printf issues resolved

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 1]
```diff
@@ -476,10 +476,24 @@ public function __clone()
      */
     public function __toString()
     {
+        $protocol = '';
+        try {
+            $protocol = $this->server->get('SERVER_PROTOCOL');
+        } catch (\InvalidArgumentException $mallformedPath) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: We merged a PR where all caught exceptions were renamed to $e. So should be this one.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 2]
```diff
@@ -476,10 +476,24 @@ public function __clone()
...
+            trigger_error($mallformedPath->getMessage(), E_USER_ERROR);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: This does replace a fatal error by an other one with a better error message isn't it?

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: Hum, in fact, the dev error handler will throw an exception when this error will be raised. This means end up with the same behavior as before (a fatal error because __toString throws).
In prod, the error message will be a bit better though, but that's a lot of code for a very edge case solution I believe.

Well, looking at the code that throws the exception (in ParameterBag), the only way to trigger it is by using a malformed $path. But here, $path = 'SERVER_PROTOCOL'. There is no way this exception can happen.

File: src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php [hunk 3]
```diff
@@ -63,8 +63,12 @@ public function __construct(array $headers = array())
...
     {
         $cookies = '';
-        foreach ($this->getCookies() as $cookie) {
-            $cookies .= 'Set-Cookie: '.$cookie."\r\n";
+        try {
+            foreach ($this->getCookies() as $cookie) {
+                $cookies .= 'Set-Cookie: '.$cookie."\r\n";
+            }
+        } catch (\InvalidArgumentException $wrongFormat) {
+            trigger_error($wrongFormat->getMessage(), E_USER_ERROR);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php]: Same here: no exception is going to be thrown, the format is right and can't be wrong.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 4]
```diff
@@ -476,10 +476,24 @@ public function __clone()
...
+        }
+
+        $content = '';
+        try {
+            $content = $this->getContent();
+        } catch (\LogicException $usageRestrictions) {
+            trigger_error($usageRestrictions->getMessage(), E_USER_ERROR);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: This case can not happen also, because the exception is throw only when `true` given to getContent()

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 2 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Request.php]: This is really edge case which will not happen until `get` is unchanged.

In general `trigger_error` shall spam php error log at least (ideally will be noticed).
In cases with raising exceptions by error handler - yes, it will be still fatal.

I'd propose to keep this, debugging this kind of issues normally takes too much time.

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Request.php]: Emm, why? Code shall be self-explaining.

I could then rename $protocol to $p - it's clear from context what happens...

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: I'm wrong on this one: it can happen if getContent() has be *previously* called with `true`.
I'd be OK to keep this one, but :-1: for the other.

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: Still, please rename $usageRestrictions to $e

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: If you have xdebug or the symfony_debug extensions, then you can get the stack trace of fatal error. This is how such issue are easily debugged. No need to keep "long distance relationships" within the code base.

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: See #14937

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 2 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Request.php]: Ok, I saw the point, will revert these changes.

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 4 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Request.php]: Ok, `getContent` try-catch wrapper will remain, `->$e` will be done.

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Request.php]: I see, ths for the link.

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @kalessil.
